pgjdbc/src/main/java/org/postgresql/core/ParameterContext.java

```diff
@@ -0,0 +1,192 @@
+      placeholderPositions = new ArrayList<>();
+    }
+
+    placeholderPositions.add(position);
+    int bindIndex = placeholderPositions.size() - 1;
+
+    if (placeholderAtPosition == null) {
+      placeholderAtPosition = new ArrayList<>();
+    }
+
+    placeholderAtPosition.add(bindIndex);
+
+    return bindIndex + 1;
+  }
+
+  public boolean hasParameters() {
+    return placeholderPositions != null && !placeholderPositions.isEmpty();
+  }
+
+  public boolean hasNamedParameters() {
+    return placeholderNames != null && !placeholderNames.isEmpty();
+  }
+
+  /**
+   * @param i 0-indexed position in the order of first appearance
+   * @return The name of the placeholder at this backend parameter position
+   */
+  public String getPlaceholderName(int i) {
+    if (placeholderNames == null) {
+      throw new IllegalStateException("Call hasNamedParameters() first.");
+    }
+    return placeholderNames.get(i);
+  }
+
+  /**
+   * @param i 0-indexed position in the order of appearance
+   * @return The position of the placeholder in the SQL text for this placeholder index
+   */
+  public int getPlaceholderPosition(int i) {
+    if (placeholderPositions == null) {
+      throw new IllegalStateException("Call hasParameters() first.");
+    }
+    return placeholderPositions.get(i);
+  }
+
+  /**
+   * @param i 0-indexed position in the order of appearance
+   * @return The position of the placeholder in the order of first appearance of each placeholder
+   */
+  public int getPlaceholderAtPosition(int i) {
+    if (placeholderAtPosition == null) {
+      throw new IllegalStateException("Call hasParameters() first.");
+    }
+    return placeholderAtPosition.get(i);
+  }
+
+  public int getLastPlaceholderPosition() {
+    if (placeholderPositions == null) {
+      throw new IllegalStateException("Call hasParameters() first.");
+    }
+    return placeholderPositions.get(placeholderPositions.size() - 1);
+  }
+
+  /**
+   * Adds a named parameter to this ParameterContext.
+   * Once a named Parameter have been added all subsequent parameters must be named.
+   * Using named parameters enable reuse of the same parameters in several locations of the SQL
+   * text. The parameters only have to be sent to the PostgreSQL backend once per name specified.
+   * The values will be sent in the order of the first appearance of their placeholder.
+   *
+   * @param position in the SQL text where the parser captured the placeholder.
+   * @param bindName is the placeholder name captured by the parser.
+   * @return 1-indexed position in the order of first appearance of named parameters
+   * @throws SQLException if positional and named parameters are mixed.
+   */
+  public int addNamedParameter(int position, String bindName) throws SQLException {
+    if (bindStyle == BindStyle.POSITIONAL) {
+      throw new SQLException("Positional and named parameters cannot be combined! Saw "
+          + "positional parameter first.");
+    }
```

Comment:
How about centralizing this and the positional parameter `bindStyle` validation and checking for the affirmative?

```java
private void checkAndSetBindStyle(BindStyle bindStyle) {
  if (this.bindStyle == null) {
    this.bindStyle = bindStyle;
  } else if (this.bindStyle != bindStyle) {
    throw new SQLException("Multiple bind styles cannot be combined. Saw " + this.bindStyle  + " first but attempting to also use: " + bindStyle);
  }
}

public int addPositionalParameter(int position, String bindName) throws SQLException {
    checkAndSetBindStyle(BindStyle.POSITIONAL);
    // ...
}

public int addNamedParameter(int position, String bindName) throws SQLException {
    checkAndSetBindStyle(BindStyle.NAMED);
    // ...
}
```

This also has the advantage of supporting additional bind styles without changing anything at these locations.

Could also add a `BindStyle.NONE` for the empty context.

Reply:
Yes, this is a good solution. I did not include NONE, as I went with overriding the add-methods.